### PR TITLE
feat(usb): Add USB polling interval patch

### PR DESCRIPTION
**Pull request by arthur-proglove**
This patch allows userland to configure the HID USB polling interval
via ConfigFS

echo X > /sys/kernel/config/usb_gadget/gateway/functions/hid.usb0/interval

### CODE REVIEW

File: patches/usb-gadget-interval-hid.patch [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+ 	hidg_fs_out_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_fs_out_ep_desc.bInterval = interval_ms;
+ 	/*
+ 	 * We can use hidg_desc struct here but we should not relay
+ 	 * that its content won't change after returning from this function.
+@@ -924,6 +914,7 @@ CONFIGFS_ATTR(f_hid_opts_, name)
+ F_HID_OPT(subclass, 8, 255);
+ F_HID_OPT(protocol, 8, 255);
+ F_HID_OPT(report_length, 16, 65535);
++F_HID_OPT(interval, 16, 65535);
```
Comment (dwolfi) [inline patches/usb-gadget-interval-hid.patch]: why 16?

File: patches/usb-gadget-interval-hid.patch [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+ 	hidg_interface_desc.bInterfaceProtocol = hidg->bInterfaceProtocol;
+ 	hidg->protocol = HID_REPORT_PROTOCOL;
++	hidg_ss_in_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_ss_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
+ 	hidg_ss_in_comp_desc.wBytesPerInterval =
+ 				cpu_to_le16(hidg->report_length);
+ 	hidg_hs_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_hs_in_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_fs_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_fs_in_ep_desc.bInterval = interval_ms;
```
Comment (dwolfi) [inline patches/usb-gadget-interval-hid.patch:104]: is this intentional that the same variable is assigned both?
- `USB_MS_TO_HS_INTERVAL(interval_ms)`
- `interval_ms`

**Review by dwolfi — APPROVED** [review]

File: patches/usb-gadget-interval-hid.patch (hunk 1 above)
Comment (arthur-proglove) [reply patches/usb-gadget-interval-hid.patch:120]: fair point, up to 255ms should be enough :-)

Comment (arthur-proglove) [reply patches/usb-gadget-interval-hid.patch:120]: In case we want more than 255ms :-)

Comment (arthur-proglove) [reply patches/usb-gadget-interval-hid.patch]: yes :-)

Comment (arthur-proglove) [reply patches/usb-gadget-interval-hid.patch]: fixed

File: patches/usb-gadget-interval-hid.patch (hunk 2 above)
Comment (arthur-proglove) [reply patches/usb-gadget-interval-hid.patch:104]: it's normal:

- Full Speed interval can be taken as it is (in ms)

```
	usb_ep_desc_binterval_fs_1_ms	= 1,
	usb_ep_desc_binterval_fs_2_ms	= 2,
	usb_ep_desc_binterval_fs_4_ms	= 4,
	usb_ep_desc_binterval_fs_8_ms	= 8,
	usb_ep_desc_binterval_fs_16_ms	= 16,
	usb_ep_desc_binterval_fs_32_ms	= 32,
```

- SuperSpeed and HighSpeed need to be converted.

```
	usb_ep_desc_binterval_hs_125_us	= 1,
	usb_ep_desc_binterval_hs_250_us	= 2,
	usb_ep_desc_binterval_hs_500_us	= 3,
	usb_ep_desc_binterval_hs_1_ms	= 4,
	usb_ep_desc_binterval_hs_2_ms	= 5,
	usb_ep_desc_binterval_hs_4_ms	= 6
```

**Review by JoeHut — APPROVED** [review]

**Pull request merged**